### PR TITLE
Yashveer/baseline model

### PR DESCRIPTION
**Pull request by yashveersinghsohi**
This branch aims to implement the first version of a model used to predict team performance using the "featurized" datasets  in the output directory.

We allow users to select a target for each of the datasets, and then build a auto hyperparameter optimized XGB regression model for the specified target.

### CODE REVIEW

**Comment by yashveersinghsohi** [comment]
Made some of the proposed changes from #116 . Primarily around adding in metrics and diagnostics for a range of models other than XGBoost, and made the pipeline more generalizable (custom number of trials instead of the hardcoded 15, and ability to pick the number of folds in k-fold cross validation).

**Comment by xehu** [comment]
Just testing through everything and leaving a few comments!

It seems like the stats for linear are really, really weird:
Screenshot 2023-05-29 at 8 33 50 PM

This might get back to the question we'd discussed, about using train/test instead of k-fold cross validation --- since the model stats seem really weird, but it's hard to debug since it's not super interpretable. 

Additionally, when running 'optimize_model' on xgb, I'm getting silent failures:
Screenshot 2023-05-29 at 9 01 05 PM

And it seems that optimization only works for `xgb` right now --- there's really nothing to optimize in a linear model (it's a closed-form solution), but it fails loudly when called on anything other than xgb:
Screenshot 2023-05-29 at 9 07 54 PM

**Comment by yashveersinghsohi** [comment]
Thanks for these comments Emily. The optimization part is not built for other models yet. That is one of the action items for the current batch of commit going through. 

Once that stuff is merged in, I will mark this branch ready for review. In the meantime feel free to take a pause on reviewing this PR (I am basically adding all the discussed action items for this week in the branch itself).

**Comment by xehu** [comment]
@yashveersinghsohi ack! My bad for not pulling the latest edits --- I deleted my previous comment because I realized it was no longer relevant given the changes that are currently in progress.

Before I saw your most recent comment, I had pulled the latest changes (converting the k-fold cross-validation to train/test), and played around a bit; I added a test in the iPython notebook for adding a custom test set (e.g,. training on CSOP and testing on CSOP II). I also made a few modifications: 
- I collapsed the two `select_target` functions down to call a single underlying function, since they're essentially doing the same thing;
- I changed the interface slightly to require the user to separately select targets for the main dataset and the test dataset --- this is because the target in CSOP and CSOP II actually happen to have different names.
- I added slight amounts of error checking (e.g., it is acceptable for the user to provide the target as either a string or as a list).

Overall, the train/test split looks great, and I love that it's far more interpretable than the cross-validation!

I think optimization is probably still under construction --- currently all 4 model types throw an error:

Screenshot 2023-05-30 at 12 49 58 AM

I'm going to switch off this branch now to look at other PR's, and leave you be! I had told you earlier that I was going to review it and try to push it in, but given hectic travel and shaky wi-fi access, I wasn't able to do it sooner, and I hadn't realized that now it's under progress / draft status again. Sorry about that; great work so far!!

**Comment by xehu** [comment]
@yashveersinghsohi -- just a quick FYI; Shruti's first feature just went in, which adds 20 features to all the datasets. It might be a good idea to pull in main (as the data are now stale); would be cool to see what the models look like with the new politeness features!

Additionally, related to our idea of creating "bags" of related features, all 20 features that just went in are from the same function --- they're a collection of features that indicate "politeness." Some of them are also directly or indirectly redundant with existing features; something that I'm looking into exploring. For example, positive/negative words, first person, and hedge are all included as politeness markers, but all of these are also things that already exist elsewhere in the data.

**Comment by yashveersinghsohi** [comment]
Thank you for the heads up @xehu!

I merged in her updates successfully, and on the CSOP datasets (train/val - CSOP, test CSOP-II) we get a 2% increase in performance in the validation and test sets (without any optimization). So we are in the right direction with these new features.

**Comment by xehu** [comment]
@yashveersinghsohi omg amazing -- I am posting this update in the main channel because that is super cool

**Comment by yashveersinghsohi** [comment]
Hey @xehu.

I am marking this ready to review. I have added in some notes (in #116 ) about what to expect in this PR, and what tasks from #116 can be designated to future PRs and branches. Do let me know what you feel about the comments and the code changes here overall.

Thanks again for being so thorough and active in this review process!

**Review by xehu — APPROVED** [review]
looks super great!! :)
